csrc/multidevice/cuda_p2p.cpp

```diff
@@ -0,0 +1,70 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <cuda_utils.h>
+#include <multidevice/cuda_p2p.h>
+
+namespace nvfuser {
+
+namespace getZcopy {
+
+void RecvPost(const P2pIpcHandle& ipc_handles, int64_t count, CUstream stream) {
+  // wait for sender to be ready
+  NVFUSER_CUDA_SAFE_CALL(cuStreamWaitValue32(
+      stream,
+      reinterpret_cast<CUdeviceptr>(ipc_handles.local().semaphore()),
```

Comment:
Why two semaphores (local and peer)? AFAICT, a semaphore is shared between sender and receiver so both devices see the same value. Apparently, one semaphore per P2pIpcHandle indicating NOT_READY, READY, or COMPLETE ought to be enough?

Reply:
You are right that we **could** use only one semaphore. The question would then be where to allocate it (on the sender of receiver GPU).

The idea behind why we use two semaphores here, is to make sure that `cuStreamWaitValue32` always polls a local buffer. That is considered good practice -- we always want the listener to poll a local buffer and not a remote one to avoid too much network transactions. That is why the semaphore is duplicated, one on the recv and one on the sender GPU. We pay the cost of duplicating the `cuStreamWriteValue32` calls, to update both semaphore each time, but that's a very minor drawback.

I have not run any performance benchmark on this, though.